### PR TITLE
Fix #17505 - Retry logic for Gemini should support multiple transactions

### PR DESCRIPTION
**Pull request by jumde**
Resolves https://github.com/brave/brave-browser/issues/17505

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Initiate two tips one after the other, verify that both tips complete after the settlement time.
2. Initiate a tip once the re-concile interval finishes, verify that both auto-contribute and tips complete after the settlement time elapses.

### CODE REVIEW

File: vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.cc [hunk 1]
```diff
@@ -101,11 +102,13 @@ void GeminiTransfer::OnTransactionStatus(const type::Result result,
                                          const int attempts,
                                          client::TransactionCallback callback) {
   if (result == type::Result::LEDGER_OK) {
+    retry_timer_.erase(id);
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.cc:105]: Should we just always call `erase` as the first thing that this callback does?

Comment (jumde) [reply vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.cc]: resolved!

**Review by emerick — APPROVED** [review]
LGTM

File: vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.h [hunk 2]
```diff
@@ -50,7 +50,7 @@ class GeminiTransfer {
                                    const int attempts,
                                    client::TransactionCallback callback);
 
-  base::OneShotTimer retry_timer_;
+  std::map<std::string, std::unique_ptr<base::OneShotTimer>> retry_timer_;
```
Comment (szilardszaloki) [inline vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.h:53]: Any particular reason to use `std::unique_ptr<>`?

Comment (jumde) [reply vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.h:53]: It's easy to manage scope with a unique ptr, I found this convention used in ads_service_impl as well - https://github.com/brave/brave-core/blob/master/components/brave_ads/browser/ads_service_impl.cc#L1926

Let me know if you have any concerns.

Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.h:53]: Yeah, I'm not quite sure why it's used there either. 🙂
In a `std::map<std::string, base::OneShotTimer>`, the lifetime of  `base::OneShotTimer`s is already managed by the red-black tree implementation of the `std::map<>`. Storing `std::unique_ptr<base::OneShotTimer>`s instead results in unnecessary heap allocations/deallocations.

Comment (jumde) [reply vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.h:53]: ah! this is helpful, I'll update. Thanks @szilardszaloki

**Review by szilardszaloki — APPROVED** [review]

**Review by zenparsing — APPROVED** [review]

**Comment by LaurenWags** [comment]
Verification in progress with

```
Brave | 1.30.38 Chromium: 93.0.4577.51 (Official Build) nightly (x86_64)
-- | --
Revision | 762d21050e2da59930c784c09b134d0b0b148188-refs/branch-heads/4577@{#915}
OS | macOS Version 11.5.2 (Build 20G95)
```

<details>
<summary>Reproduced the issue with 1.29.66</summary>

Tip 1 is 1 BAT:

Example | Example
----- | -----
Screen Shot 2021-08-20 at 2 43 46 PM | Screen Shot 2021-08-20 at 2 44 20 PM

<details>
<summary>Logs</summary>

```
[3490:259:0820/144407.767694:VERBOSE1:contribution.cc(362)] Creating contribution for wallet type gemini (amount: 1, type: ONE_TIME_TIP)
[3490:259:0820/144407.771432:VERBOSE1:contribution.cc(91)] Queue timer set for 15 s
[3490:259:0820/144407.772188:VERBOSE1:database_server_publisher_banner.cc(142)] Server publisher banner not found
[3490:259:0820/144407.773892:INFO:post_transaction_gemini.cc(103)] Initiating gemini transaction to: 60e86f02-81f8-40fd-86f6-c2d6be298b37for 0.95
[3490:259:0820/144407.773965:VERBOSE5:ledger_impl.cc(119)] 
[ REQUEST ]
> URL: https://api.qa6.aurora7.net/v1/payments/pay
> Method: POST
> Content Type: application/json; charset=utf-8
[3490:259:0820/144407.975539:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnRequest ]
> Url: https://api.qa6.aurora7.net/v1/payments/pay
> Result: Success
> HTTP Code: 200
> Body: {"result":"OK","tx_ref":"f942edab-4a35-493b-aec5-3a03bfc215c8","amount":0.95,"currency":"BAT","destination":"60e86f02-81f8-40fd-86f6-c2d6be298b37","status":"Pending"}
[3490:259:0820/144407.975626:VERBOSE9:logging_util.cc(137)] 
[ RESPONSE HEADERS ]
> Url: https://api.qa6.aurora7.net/v1/payments/pay
[3490:259:0820/144407.976526:INFO:post_transaction_gemini.cc(116)] Gemini transaction successful
[3490:259:0820/144407.976590:VERBOSE1:post_transaction_gemini.cc(129)] Transfer Pending
[3490:259:0820/144407.976636:VERBOSE1:gemini_transfer.cc(75)] Will fetch transaction status after 3 minutes
```

</details>

Tip 2 is for 3 BAT:

Example | Example
----- | -----
Screen Shot 2021-08-20 at 2 43 57 PM | Screen Shot 2021-08-20 at 2 44 28 PM

<details>
<summary>Logs</summary>

```
[3490:259:0820/144423.446290:VERBOSE1:contribution.cc(362)] Creating contribution for wallet type gemini (amount: 3, type: ONE_TIME_TIP)
[3490:259:0820/144423.450178:VERBOSE1:contribution.cc(91)] Queue timer set for 7 s
[3490:259:0820/144423.450883:VERBOSE1:database_server_publisher_banner.cc(142)] Server publisher banner not found
[3490:259:0820/144423.452215:INFO:post_transaction_gemini.cc(103)] Initiating gemini transaction to: 60ec69b1-f02c-467e-9247-8c174c49eb95for 2.85
[3490:259:0820/144423.452272:VERBOSE5:ledger_impl.cc(119)] 
[ REQUEST ]
> URL: https://api.qa6.aurora7.net/v1/payments/pay
> Method: POST
> Content Type: application/json; charset=utf-8
[3490:259:0820/144423.633701:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnRequest ]
> Url: https://api.qa6.aurora7.net/v1/payments/pay
> Result: Success
> HTTP Code: 200
> Body: {"result":"OK","tx_ref":"31d0a168-f609-4a7b-a90f-037dc569ba02","amount":2.85,"currency":"BAT","destination":"60ec69b1-f02c-467e-9247-8c174c49eb95","status":"Pending"}
[3490:259:0820/144423.633809:VERBOSE9:logging_util.cc(137)] 
[ RESPONSE HEADERS ]
> Url: https://api.qa6.aurora7.net/v1/payments/pay
[3490:259:0820/144423.634929:INFO:post_transaction_gemini.cc(116)] Gemini transaction successful
[3490:259:0820/144423.635014:VERBOSE1:post_transaction_gemini.cc(129)] Transfer Pending
[3490:259:0820/144423.635093:VERBOSE1:gemini_transfer.cc(75)] Will fetch transaction status after 3 minutes
```

```
[3490:259:0820/144723.637784:VERBOSE1:gemini_transfer.cc(86)] Fetching transaction status 1 time
[3490:259:0820/144723.638427:VERBOSE5:ledger_impl.cc(119)] 
[ REQUEST ]
> URL: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/31d0a168-f609-4a7b-a90f-037dc569ba02
> Method: GET
> Content Type: application/json; charset=utf-8
[3490:259:0820/144724.133057:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnRequest ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/31d0a168-f609-4a7b-a90f-037dc569ba02
> Result: Success
> HTTP Code: 200
> Body: {"tx_ref":"31d0a168-f609-4a7b-a90f-037dc569ba02","amount":2.85,"currency":"BAT","destination":"60ec69b1-f02c-467e-9247-8c174c49eb95","status":"Pending"}
[3490:259:0820/144724.133158:VERBOSE9:logging_util.cc(137)] 
[ RESPONSE HEADERS ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/31d0a168-f609-4a7b-a90f-037dc569ba02
[3490:259:0820/144724.133213:VERBOSE1:get_transaction_gemini.cc(88)] Transfer Status: Pending
[3490:259:0820/144724.133263:VERBOSE1:gemini_transfer.cc(75)] Will fetch transaction status after 6 minutes
```

```
[3490:259:0820/145324.138485:VERBOSE1:gemini_transfer.cc(86)] Fetching transaction status 2 time
[3490:259:0820/145324.139698:VERBOSE5:ledger_impl.cc(119)] 
[ REQUEST ]
> URL: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/31d0a168-f609-4a7b-a90f-037dc569ba02
> Method: GET
> Content Type: application/json; charset=utf-8
[3490:259:0820/145324.314828:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnRequest ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/31d0a168-f609-4a7b-a90f-037dc569ba02
> Result: Success
> HTTP Code: 200
> Body: {"tx_ref":"31d0a168-f609-4a7b-a90f-037dc569ba02","amount":2.85,"currency":"BAT","destination":"60ec69b1-f02c-467e-9247-8c174c49eb95","status":"Completed","timestampms":1629485590210}
[3490:259:0820/145324.314939:VERBOSE9:logging_util.cc(137)] 
[ RESPONSE HEADERS ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/31d0a168-f609-4a7b-a90f-037dc569ba02
[3490:259:0820/145324.314997:VERBOSE1:get_transaction_gemini.cc(88)] Transfer Status: Completed
[3490:259:0820/145324.315047:VERBOSE1:gemini.cc(212)] Gemini transfer fee timer set for 6 s
[3490:259:0820/145324.333446:VERBOSE1:uphold.cc(114)] Uphold wallet is not VERIFIED.
[3490:259:0820/145324.334095:VERBOSE5:ledger_impl.cc(119)] 
[ REQUEST ]
> URL: https://api.qa6.aurora7.net/v1/balances
> Method: POST
[3490:259:0820/145324.496182:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnRequest ]
> Url: https://api.qa6.aurora7.net/v1/balances
> Result: Success
> HTTP Code: 200
> Body: [{"type":"exchange","currency":"BAT","amount":"10021.5","available":"10021.5","availableForWithdrawal":"10021.5"}]
[3490:259:0820/145324.496289:VERBOSE9:logging_util.cc(137)] 
[ RESPONSE HEADERS ]
> Url: https://api.qa6.aurora7.net/v1/balances
[3490:259:0820/145328.680683:VERBOSE1:ads_impl.cc(209)] Browser state changed to idle
[3490:259:0820/145330.323827:INFO:post_transaction_gemini.cc(103)] Initiating gemini transaction to: 60edbc1b-eaa7-4aa0-afb8-a87a89634c55for 0.15
[3490:259:0820/145330.323942:VERBOSE5:ledger_impl.cc(119)] 
[ REQUEST ]
> URL: https://api.qa6.aurora7.net/v1/payments/pay
> Method: POST
> Content Type: application/json; charset=utf-8
[3490:259:0820/145330.513582:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnRequest ]
> Url: https://api.qa6.aurora7.net/v1/payments/pay
> Result: Success
> HTTP Code: 200
> Body: {"result":"OK","tx_ref":"99393aaf-6bcc-4b78-8f89-0d25d9f7dc91","amount":0.15,"currency":"BAT","destination":"60edbc1b-eaa7-4aa0-afb8-a87a89634c55","status":"Pending"}
[3490:259:0820/145330.513683:VERBOSE9:logging_util.cc(137)] 
[ RESPONSE HEADERS ]
> Url: https://api.qa6.aurora7.net/v1/payments/pay
[3490:259:0820/145330.514777:INFO:post_transaction_gemini.cc(116)] Gemini transaction successful
[3490:259:0820/145330.514859:VERBOSE1:post_transaction_gemini.cc(129)] Transfer Pending
[3490:259:0820/145330.514915:VERBOSE1:gemini_transfer.cc(75)] Will fetch transaction status after 3 minutes
```

```
[3490:259:0820/145630.517349:VERBOSE1:gemini_transfer.cc(86)] Fetching transaction status 1 time
[3490:259:0820/145630.518550:VERBOSE5:ledger_impl.cc(119)] 
[ REQUEST ]
> URL: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/99393aaf-6bcc-4b78-8f89-0d25d9f7dc91
> Method: GET
> Content Type: application/json; charset=utf-8
[3490:259:0820/145630.688970:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnRequest ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/99393aaf-6bcc-4b78-8f89-0d25d9f7dc91
> Result: Success
> HTTP Code: 200
> Body: {"tx_ref":"99393aaf-6bcc-4b78-8f89-0d25d9f7dc91","amount":0.15,"currency":"BAT","destination":"60edbc1b-eaa7-4aa0-afb8-a87a89634c55","status":"Pending"}
[3490:259:0820/145630.689077:VERBOSE9:logging_util.cc(137)] 
[ RESPONSE HEADERS ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/99393aaf-6bcc-4b78-8f89-0d25d9f7dc91

> Header content-length: 152
> Header content-type: application/json
> Header date: Fri, 20 Aug 2021 18:56:30 GMT
> Header server: nginx
> Header vary: Origin
[3490:259:0820/145630.689135:VERBOSE1:get_transaction_gemini.cc(88)] Transfer Status: Pending
[3490:259:0820/145630.689187:VERBOSE1:gemini_transfer.cc(75)] Will fetch transaction status after 6 minutes
```

```
[3490:259:0820/150230.695610:VERBOSE1:gemini_transfer.cc(86)] Fetching transaction status 2 time
[3490:259:0820/150230.696795:VERBOSE5:ledger_impl.cc(119)] 
[ REQUEST ]
> URL: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/99393aaf-6bcc-4b78-8f89-0d25d9f7dc91
> Method: GET
> Content Type: application/json; charset=utf-8
[3490:259:0820/150230.871108:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnRequest ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/99393aaf-6bcc-4b78-8f89-0d25d9f7dc91
> Result: Success
> HTTP Code: 200
> Body: {"tx_ref":"99393aaf-6bcc-4b78-8f89-0d25d9f7dc91","amount":0.15,"currency":"BAT","destination":"60edbc1b-eaa7-4aa0-afb8-a87a89634c55","status":"Pending"}
[3490:259:0820/150230.871216:VERBOSE9:logging_util.cc(137)] 
[ RESPONSE HEADERS ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/99393aaf-6bcc-4b78-8f89-0d25d9f7dc91

> Header content-length: 152
> Header content-type: application/json
> Header date: Fri, 20 Aug 2021 19:02:30 GMT
> Header server: nginx
> Header vary: Origin
[3490:259:0820/150230.871276:VERBOSE1:get_transaction_gemini.cc(88)] Transfer Status: Pending
[3490:259:0820/150230.871329:VERBOSE1:gemini_transfer.cc(75)] Will fetch transaction status after 9 minutes
```

```
[3490:259:0820/151130.878758:VERBOSE1:gemini_transfer.cc(86)] Fetching transaction status 3 time
[3490:259:0820/151130.879801:VERBOSE5:ledger_impl.cc(119)] 
[ REQUEST ]
> URL: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/99393aaf-6bcc-4b78-8f89-0d25d9f7dc91
> Method: GET
> Content Type: application/json; charset=utf-8
[3490:259:0820/151131.051688:VERBOSE6:logging_util.cc(136)] 
[ RESPONSE - OnRequest ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/99393aaf-6bcc-4b78-8f89-0d25d9f7dc91
> Result: Success
> HTTP Code: 200
> Body: {"tx_ref":"99393aaf-6bcc-4b78-8f89-0d25d9f7dc91","amount":0.15,"currency":"BAT","destination":"60edbc1b-eaa7-4aa0-afb8-a87a89634c55","status":"Completed","timestampms":1629486190024}
[3490:259:0820/151131.051774:VERBOSE9:logging_util.cc(137)] 
[ RESPONSE HEADERS ]
> Url: https://api.qa6.aurora7.net/v1/payment/60df8caa-795b-4eef-a8bf-585e3799d454/99393aaf-6bcc-4b78-8f89-0d25d9f7dc91
[3490:259:0820/151131.051821:VERBOSE1:get_transaction_gemini.cc(88)] Transfer Status: Completed
```

</details>

Saw that Tip 1 (1 BAT) did not complete while Tip 2 (3 BAT) did complete:

Example | Example | Example
----- | ----- | -----
Screen Shot 2021-08-20 at 3 12 49 PM | Screen Shot 2021-08-20 at 3 12 58 PM | Screen Shot 2021-08-20 at 3 13 07 PM


</details>

**Pull request merged**